src/helpers/SparseCalldataSegmentLib.sol

```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.25;
+
+library SparseCalldataSegmentLib {
+    /// @notice Splits out a segment of calldata, sparsely-packed
+    /// @param source The calldata to extract the segment from
+    /// @return segment The extracted segment
+    /// @return remainder The remaining calldata
+    function getNextSegment(bytes calldata source)
+        internal
+        pure
+        returns (bytes calldata segment, bytes calldata remainder)
+    {
+        // The first 8 bytes hold the length of the segment, excluding the index.
+        uint64 length = uint64(bytes8(source[:8]));
```

Comment:
We might also want to consider edge cases like insufficient length

Reply:
Thankfully, solidity will revert in the array slice step (in this case, `source[:8]`) if the length is insufficient: https://docs.soliditylang.org/en/v0.8.26/types.html#array-slices

Reply:
Slight tangent, but why reserve 8 bytes for the length? Could we use less bytes?

Reply:
Possibly yes, we could use fewer. 4 bytes would give us a max length of ~4 billion, or 4 gigabytes per segment, which seems sufficient.

Reply:
Changed to 4 bytes